File: plugin/src/test/resources/golden/rust/address.rs
use serde_json::{Value, json};

#[derive(Debug, Clone, Default)]
pub struct Address {
    pub street: String,
    pub city: String,
    pub state: String,
    pub zip: i32,
}

impl Address {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn street(&self) -> &str {
        &self.street
    }

    pub fn set_street(&mut self, value: String) {
        self.street = value;
    }

    pub fn city(&self) -> &str {
        &self.city
    }

    pub fn set_city(&mut self, value: String) {
        self.city = value;
    }

    pub fn state(&self) -> &str {
        &self.state
    }

    pub fn set_state(&mut self, value: String) {
        self.state = value;
    }

    pub fn zip(&self) -> i32 {
        self.zip
    }

    pub fn set_zip(&mut self, value: i32) {
        self.zip = value;
    }

    pub fn serialize(&self) -> Value {
        let mut arr: Vec<Value> = Vec::new();
        arr.push(json!(self.street));
        arr.push(json!(self.city));
        arr.push(json!(self.state));
        arr.push(json!(self.zip));
        Value::Array(arr)
    }

    pub fn to_json_string(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(&self.serialize())
    }

    pub fn to_json_bytes(&self) -> Result<Vec<u8>, serde_json::Error> {
        Ok(self.to_json_string()?.into_bytes())
    }

    pub fn deserialize(arr: &Value) -> Result<Self, String> {
        let arr = arr.as_array().ok_or_else(|| "expected JSON array".to_string())?;
        let size = arr.len();
        let mut obj = Address::default();
        if size > 0 && !arr[0].is_null() {
            obj.street = arr[0].as_str().unwrap_or("").to_string();
        }
        if size > 1 && !arr[1].is_null() {
            obj.city = arr[1].as_str().unwrap_or("").to_string();
        }
        if size > 2 && !arr[2].is_null() {
            obj.state = arr[2].as_str().unwrap_or("").to_string();
        }
        if size > 3 && !arr[3].is_null() {
            obj.zip = arr[3].as_i64().unwrap_or(0) as i32;
        }
        Ok(obj)
    }

    pub fn from_json_string(json: &str) -> Result<Self, String> {
        let value: Value = serde_json::from_str(json).map_err(|e| e.to_string())?;
        Self::deserialize(&value)
    }

    pub fn from_json_bytes(bytes: &[u8]) -> Result<Self, String> {
        let json = std::str::from_utf8(bytes).map_err(|e| e.to_string())?;
        Self::from_json_string(json)
    }
}

impl std::fmt::Display for Address {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Address{{street: {:?}, city: {:?}, state: {:?}, zip: {:?}}}", self.street, self.city, self.state, self.zip)
    }
}

File: plugin/src/test/resources/golden/zig/address.zig
const std = @import("std");
const json = std.json;

pub const Address = struct {
    street: []const u8 = "",
    city: []const u8 = "",
    state: []const u8 = "",
    zip: i32 = 0,

    pub fn serializeToValue(self: *const Address, allocator: std.mem.Allocator) !json.Value {
        var arr = try allocator.alloc(json.Value, 4);
        arr[0] = json.Value{ .string = self.street };
        arr[1] = json.Value{ .string = self.city };
        arr[2] = json.Value{ .string = self.state };
        arr[3] = json.Value{ .integer = @as(i64, @intCast(self.zip)) };
        return json.Value{ .array = json.Array.fromOwnedSlice(allocator, arr) };
    }

    pub fn serialize(self: *const Address, allocator: std.mem.Allocator) ![]u8 {
        const value = try self.serializeToValue(allocator);
        var string = std.ArrayList(u8).init(allocator);
        defer string.deinit();
        try value.jsonStringify(.{}, string.writer());
        return try string.toOwnedSlice();
    }

    pub fn deserializeFromValue(value: json.Value, allocator: std.mem.Allocator) !Address {
        const arr = value.array.items;
        const size = arr.len;
        var obj: Address = undefined;
        obj.street = "";
        obj.city = "";
        obj.state = "";
        obj.zip = 0;

        if (size > 0 and arr[0] != .null) {
            obj.street = try allocator.dupe(u8, arr[0].string);
        }
        if (size > 1 and arr[1] != .null) {
            obj.city = try allocator.dupe(u8, arr[1].string);
        }
        if (size > 2 and arr[2] != .null) {
            obj.state = try allocator.dupe(u8, arr[2].string);
        }
        if (size > 3 and arr[3] != .null) {
            obj.zip = @as(i32, @intCast(arr[3].integer));
        }
        return obj;
    }

    pub fn deserialize(data: []const u8, allocator: std.mem.Allocator) !Address {
        const parsed = try std.json.parseFromSlice(json.Value, allocator, data, .{});
        defer parsed.deinit();
        return try deserializeFromValue(parsed.value, allocator);
    }

    pub fn deinit(self: *Address, allocator: std.mem.Allocator) void {
        if (self.street.len > 0) {
            allocator.free(self.street);
        }
        if (self.city.len > 0) {
            allocator.free(self.city);
        }
        if (self.state.len > 0) {
            allocator.free(self.state);
        }
    }
};
